chalice/deploy/packager.py

```diff
@@ -208,3 +216,373 @@ def _add_chalice_lib_if_needed(self, project_dir, zip_fileobj):
+        egg_info_dir = self._osutils.joinpath(package_dir, 'egg-info')
+        self._osutils.makedirs(egg_info_dir)
+        cmd += ['--egg-base', 'egg-info']
+        p = subprocess.Popen(cmd, cwd=package_dir,
+                             stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+        p.communicate()
+        self._osutils.joinpath(egg_info_dir)
+        info_contents = self._osutils.get_directory_contents(egg_info_dir)
+        assert len(info_contents) == 1
+        pkg_info_path = self._osutils.joinpath(
+            egg_info_dir, info_contents[0], 'PKG-INFO')
+        return pkg_info_path
+
+    def _unpack_sdist_into_dir(self, sdist_path, unpack_dir):
+        # type: (str, str) -> str
+        if sdist_path.endswith('zip'):
+            with zipfile.ZipFile(sdist_path, 'r') as z:
+                z.extractall(unpack_dir)
+        elif sdist_path.endswith('.tar.gz'):
+            with tarfile.open(sdist_path, 'r:gz') as tar:
+                tar.extractall(unpack_dir)
+        else:
+            raise InvalidSourceDistributionNameError(sdist_path)
+        # There should only be one directory unpacked.
+        contents = self._osutils.get_directory_contents(unpack_dir)
+        assert len(contents) == 1
+        return self._osutils.joinpath(unpack_dir, contents[0])
+
+    def get_package_name_and_version(self, directory, filename):
+        # type: (str, str) -> Tuple[str, str]
+        sdist_path = self._osutils.joinpath(directory, filename)
+        with self._osutils.tempdir() as tempdir:
+            package_dir = self._unpack_sdist_into_dir(sdist_path, tempdir)
+            pkg_info_filepath = self._generate_egg_info(package_dir)
+            metadata = self._parse_pkg_info_file(pkg_info_filepath)
+            name = metadata['Name']
+            version = metadata['Version']
+        return name, version
+
+
+class PipWrapper(object):
+    def main(self, args):
+        # type: (List[str]) -> Tuple[Optional[bytes], Optional[bytes]]
+        raise NotImplementedError('PipWrapper.main')
+
+
+class SubprocessPip(PipWrapper):
+    """Wrapper around calling pip through a subprocess."""
+    def main(self, args):
+        # type: (List[str]) -> Tuple[Optional[bytes], Optional[bytes]]
+        python_exe = sys.executable
+        invoke_pip = [python_exe, '-m', 'pip']
+        invoke_pip.extend(args)
+        p = subprocess.Popen(invoke_pip,
+                             stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+        out, err = p.communicate()
+        return out, err
+
+
+class PipRunner(object):
+    """Wrapper around pip calls used by chalice."""
+
+    def __init__(self, pip):
+        # type: (PipWrapper) -> None
+        self._wrapped_pip = pip
+
+    def _execute(self, command, args):
+        # type: (str, List[str]) -> None
+        """Execute a pip command with the given arguments."""
+        main_args = [command] + args
+        _, err = self._wrapped_pip.main(main_args)
+        if err:
```

Comment:
Yes it will fail if it fails to get packages, which is not uncommon for smaller simple pure python packages. That is too be expected and not an error. All I really care about is if pip throws us a stack trace of some kind via stderr, and the rc does not vary for that so I can't sense what kind of error it was without parsing stderr.

I talked to @dstufft about this and if I remember correctly this was what we arrived at as the best solution for catching these types of issues.